### DISCUSSION TITLE
How do you debug the app?

### DISCUSSION THREAD

**Discussion by cypherbits** [Eugeny/tabby#4210] [Q&A]
I'm almost new to nodejs, angular, electron etc...

I wanted to contribute with a few things but I can't get my breakpoints to work. I'm using Webstorm and tried VSCode but nothing... VSCode I think it says "unbound breakpoint". 

Tried many things already.

Question 2: do you really have to rebuild everything if you change just a line of code?? 

Thanks.

**Comment by Eugeny** [comment]
Hi, there are two main ways to debug stuff:
* Use the embedded Chrome Devtools (Settings -> Application -> Devtools) - it includes a debugger that "just works".
* Use VSCode with the `launch.json` from the repo. It provides a composite launch configuration that starts the main process and also attaches a debugger to the Electron renderer process.

You don't have to rebuild everything from scratch - use `yarn run watch` to start an incremental build that will live-rebuild code as you change it. You can then reload the Tabby window with Ctrl-R / Cmd-R

**Reply by cypherbits** [reply]
Ok, many thanks. I will try it.
